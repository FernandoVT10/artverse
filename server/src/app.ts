import express from "express";
import bodyParser from "body-parser";
import routes from "./routes";
import errorHandler from "@middlewares/errorHandler";

import { PUBLIC_DIR } from "./config/constants";

const app = express();

app.use(express.static(PUBLIC_DIR));
app.use(bodyParser.json());

app.use(routes);
app.use(errorHandler());

export default app;
