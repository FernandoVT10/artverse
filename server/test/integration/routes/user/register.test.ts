import bcrypt from "bcrypt";
import supertest from "supertest";
import app from "@app";

import { User } from "@models";
import { clearDB } from "@test-utils/setupDB";

import { testUsernameField, testEmailField, testPasswordField } from "./shared";

describe("Integration POST /api/users/register", () => {
  beforeEach(async () => {
    await clearDB();
  });

  const requestData = {
    username: "alex",
    email: "alex@example.com",
    password: "secret",
  };

  const request = supertest(app);
  const requestAPI = () => request.post("/api/users/register/");

  const requestAPIAndGetUser = async (requestData: any): Promise<User> => {
    await requestAPI().send(requestData);

    return (await User.findOne({
      where: { username: requestData.username },
    })) as User;
  };

  it("should return a successful response", async () => {
    const result = await requestAPI().send(requestData).expect(200);

    const { body } = result;
    expect(body.success).toBeTruthy();
  });

  it("should create the user in the database", async () => {
    const user = await requestAPIAndGetUser(requestData);

    expect(user).toMatchObject({
      username: requestData.username,
      email: requestData.email,
    });
  });

  it("should hash the user password", async () => {
    const user = await requestAPIAndGetUser(requestData);

    const hashedPassword = user.password as string;

    expect(
      await bcrypt.compare(requestData.password, hashedPassword)
    ).toBeTruthy();
  });

  describe("validation", () => {
    testUsernameField(requestAPI, requestData);
    testEmailField(requestAPI, requestData);
    testPasswordField(requestAPI, requestData);
  });
});
