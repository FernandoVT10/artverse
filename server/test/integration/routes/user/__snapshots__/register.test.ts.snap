// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration POST /api/users/register validation email field should return an error when the email already exists in the database 1`] = `
Array [
  Object {
    "field": "email",
    "message": "The email already exists",
  },
]
`;

exports[`Integration POST /api/users/register validation email field should return an error when the email is invalid 1`] = `
Array [
  Object {
    "field": "email",
    "message": "The email is invalid",
  },
]
`;

exports[`Integration POST /api/users/register validation password field should return an error when the password is empty 1`] = `
Array [
  Object {
    "field": "password",
    "message": "The password is required",
  },
]
`;

exports[`Integration POST /api/users/register validation password field should return an error when the password is not a string 1`] = `
Array [
  Object {
    "field": "password",
    "message": "The password must be a string",
  },
]
`;

exports[`Integration POST /api/users/register validation username field should return an error when the username already exists in the database 1`] = `
Array [
  Object {
    "field": "username",
    "message": "The username already exists",
  },
]
`;

exports[`Integration POST /api/users/register validation username field should return an error when the username is larger than 30 characters 1`] = `
Array [
  Object {
    "field": "username",
    "message": "The username must have 4 or more characters",
  },
]
`;

exports[`Integration POST /api/users/register validation username field should return an error when the username is not a string 1`] = `
Array [
  Object {
    "field": "username",
    "message": "The username must be a string",
  },
]
`;

exports[`Integration POST /api/users/register validation username field should return an error when the username is shorter than 4 characters 1`] = `
Array [
  Object {
    "field": "username",
    "message": "The username must have 4 or more characters",
  },
]
`;
